spec.html

```diff
@@ -12308,50 +12241,35 @@ <h1>
-            1. If _Desc_.[[Configurable]] is present and its value is *true*, return *false*.
-            1. If _Desc_.[[Enumerable]] is present and ! SameValue(_Desc_.[[Enumerable]], _current_.[[Enumerable]]) is *false*, return *false*.
-          1. If ! IsGenericDescriptor(_Desc_) is *true*, then
-            1. NOTE: No further validation is required.
-          1. Else if ! SameValue(! IsDataDescriptor(_current_), ! IsDataDescriptor(_Desc_)) is *false*, then
-            1. If _current_.[[Configurable]] is *false*, return *false*.
-            1. If IsDataDescriptor(_current_) is *true*, then
-              1. If _O_ is not *undefined*, convert the property named _P_ of object _O_ from a data property to an accessor property. Preserve the existing values of the converted property's [[Configurable]] and [[Enumerable]] attributes and set the rest of the property's attributes to their <emu-xref href="#table-default-attribute-values">default values</emu-xref>.
-            1. Else,
-              1. If _O_ is not *undefined*, convert the property named _P_ of object _O_ from an accessor property to a data property. Preserve the existing values of the converted property's [[Configurable]] and [[Enumerable]] attributes and set the rest of the property's attributes to their <emu-xref href="#table-default-attribute-values">default values</emu-xref>.
-          1. Else if IsDataDescriptor(_current_) and IsDataDescriptor(_Desc_) are both *true*, then
-            1. If _current_.[[Configurable]] is *false* and _current_.[[Writable]] is *false*, then
-              1. If _Desc_.[[Writable]] is present and _Desc_.[[Writable]] is *true*, return *false*.
-              1. If _Desc_.[[Value]] is present and SameValue(_Desc_.[[Value]], _current_.[[Value]]) is *false*, return *false*.
-              1. Return *true*.
-          1. Else,
-            1. Assert: ! IsAccessorDescriptor(_current_) and ! IsAccessorDescriptor(_Desc_) are both *true*.
-            1. If _current_.[[Configurable]] is *false*, then
-              1. If _Desc_.[[Set]] is present and SameValue(_Desc_.[[Set]], _current_.[[Set]]) is *false*, return *false*.
-              1. If _Desc_.[[Get]] is present and SameValue(_Desc_.[[Get]], _current_.[[Get]]) is *false*, return *false*.
-              1. Return *true*.
+            1. Let _unsettableFields_ be &laquo; [[Configurable]], [[Enumerable]], [[Get]], [[Set]] &raquo;.
+            1. If ! IsAccessorDescriptor(_current_) is *true*, append [[Value]] and [[Writable]] to _unsettableFields_.
+            1. If _current_.[[Writable]] is present and is *false*, append [[Value]] and [[Writable]] to _unsettableFields_.
+            1. For each field name _field_ of _unsettableFields_, do
+              1. If _Desc_.[[&lt;_field_&gt;]] is present and either _current_.[[&lt;_field_&gt;]] is not present or ! SameValue(_Desc_.[[&lt;_field_&gt;]], _current_.[[&lt;_field_&gt;]]) is *false*, return *false*.
```

Comment:
this seems like it could be shortened to something that said "if current and desc are not the same, and !(current.writable  && !Desc.writable), then return false", without using this new dynamic slot thing?

Reply:
_current_ and _Desc_ are allowed to differ, e.g. `Object.defineProperty(Object.defineProperty({}, "x", pd1={value: 1, writable: true}), "x", pd2={writable: false})`. This is specifically checking if specific **fields** of _Desc_ are attempting to introduce invalid changes.

Reply:
there's only 5 possible field names; is the dynamic loop that much of an improvement over listing each one?

Reply:
That's a matter of opinion. The big change I like here is consolidation of "attempt to set unsettable attribute" into a single test rather than having that logic dispersed. Testing each attribute individually while preserving that consolidation would look something like the below, which seems worse to me:
```
1. If _Desc_.[[Configurable]] is present and ! SameValue(_Desc_.[[Configurable]], _current_.[[Configurable]]) is *false*, return *false*.
1. If _Desc_.[[Enumerable]] is present and ! SameValue(_Desc_.[[Enumerable]], _current_.[[Enumerable]]) is *false*, return *false*.
1. If _Desc_.[[Get]] is present and either _current_.[[Get]] is not present or ! SameValue(_Desc_.[[Get]], _current_.[[Get]]) is *false*, return *false*.
1. If _Desc_.[[Set]] is present and either _current_.[[Set]] is not present or ! SameValue(_Desc_.[[Set]], _current_.[[Set]]) is *false*, return *false*.
1. If _unsettableFields_ contains [[Value]] and _Desc_.[[Value]] is present and either _current_.[[Value]] is not present or ! SameValue(_Desc_.[[Value]], _current_.[[Value]]) is *false*, return *false*.
1. If _unsettableFields_ contains [[Writable]] and _Desc_.[[Writable]] is present and either _current_.[[Writable]] is not present or ! SameValue(_Desc_.[[Writable]], _current_.[[Writable]]) is *false*, return *false*.
```

Reply:
i'd expect unconditional checks for configurable/enumerable, and then two indented branches - one for "data" and another for "accessor", one of which that checked value/writable and the other get/set. the indirection doesn't seem as readable to me.

Reply:
I also find this loop kind of hard to read. I'd prefer to be more explicit, as @ljharb suggests, I think:

```
1. If _current_.[[Configurable]] is *false*, then
  1. If ! SameValue(! IsDataDescriptor(_current_), ! IsDataDescriptor(_Desc_)) is *false*, return *false*.
  1. If _Desc_.[[Configurable]] is present and _Desc_.[[Configurable]] is *true*, return *false*.
  1. If _Desc_.[[Enumerable]] is present and ! SameValue(_Desc_.[[Enumerable]], _current_.[[Enumerable]]) is *false*, return *false*.
  1. If ! IsDataDescriptor(_current_) is *true*, then
    1. If _current_.[[Writable]] is *false*, then
      1. If _Desc_.[[Value]] is present and ! SameValue(_Desc_.[[Value]], _current_.[[Value]]) is *false*, return *false*.
      1. If _Desc_.[[Writable]] is present and _Desc_.[[Writable]] is *true*, return *false*.
  1. Else,
    1. Assert: ! IsAccessorDescriptor(_current_) and ! IsAccessorDescriptor(_Desc_) are both *true*.
    1. If _Desc_.[[Get]] is present and ! SameValue(_Desc_.[[Get]], _current_.[[Get]]) is *false*, return *false*.
    1. If _Desc_.[[Set]] is present and ! SameValue(_Desc_.[[Set]], _current_.[[Set]]) is *false*, return *false*.
```

(We should also change to description of `_current_` to be "a fully populated Property Descriptor", since it is and that makes it clear why we don't need presence checks for `_current_`.)

Reply:
> `1. If _Desc_.[[Configurable]] is present and _Desc_.[[Configurable]] is *true*, return *false*.`

Am I alone in preferring `SameValue` over explicit `*true*`/`*false*` that requires me to scan back up and find how that relates to the covering conditions? This block is logically a collection of checks verifying the preservation of attributes that must not change, which I find to be obscured by the deep structure and divergent logic across attributes.

Reply:
All the attributes in my example use SameValue except for [[Configurable]] and [[Writable]], and for both of those I find the condition to be most clearly stated as "if false, must not become true", rather than (as is the case for the other attributes) "must not change".